gradle/copy-rest-api-spec-to-global.gradle

```diff
@@ -0,0 +1,61 @@
+
+/**
+ * Copies all of the REST api specs (including core, modules, plugins, and x-pack) to common location for the REST YAML test.
+ * This plugin is expected to only be applied to applied to the root project.
+ */
+File rootBuildDir = rootProject.buildDir
+
+tasks.create("copyCoreRestSpecGlobal", Copy) {
+  from project.findProject(':rest-api-spec').projectDir
+  into rootBuildDir
+  includeEmptyDirs = false
+  include '**/src/**/rest-api-spec/api/**'
+  eachFile {
+    path = new File("rest-api-spec/api", name)
+  }
+}
+
+tasks.create("copyModulesRestSpecGlobal", Copy) {
```

Comment:
So essentially this is the stuff you ripped out of `RestIntegTestTask`. I think that's still the right thing to do, but let's instead move it to `RestTestPlugin`. So the _plugin_ sets up the specs on the classpath instead of the task constructor.

Reply:
As mentioned above, there are many usages of `RestIntegTestTask` outside of the `RestTestPlugin` , so I can't move this to `RestTestPlugin`. I'll look into moving this back to the `RestIntegTestTask` , but not sure if that will just put us back to where we started.

Reply:
Not all `RestIntegTestTasks` require the rest specs though, that's the difference. We use those task for Java tests as well. We should put the reliance on rest specs in an explicit plugin so that if you are using YAML tests and need those specs, you _must_ apply the plugin. Otherwise we are adding the core rest specs to the classpath unnecessarily for many tasks, again, negatively affecting cacheability.

Reply:
That makes sense... However, I would like to also remove the `integTest` from the `esplugin` ([here](https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy#L167)) to help enforce this. I haven't checked yet, but for any esplugin that has REST tests, but does not have YAML based rest tests would need to include the specs unnecessarily or explicitly create the `integTest` task. I think this is pretty rare, so either option probably not much of an issue.

Thoughts on removing `integTest` from the `esplugin` to enforce the inclusion of `RestTestPlugin` for projects that have Rest tests ? This is technically a breaking change for plugin devs. (I am only targeting 8.0 anyway)

Reply:
> (I am only targeting 8.0 anyway)

We'll want to backport this to `7.x` so I don't think we can safely make this change.

> I haven't checked yet, but for any esplugin that has REST tests, but does not have YAML based rest tests would need to include the specs unnecessarily or explicitly create the integTest task. I think this is pretty rare, so either option probably not much of an issue.

I don't think so. There are a number of plugins that don't use YAML tests.

I think the problem is when we say "rest tests" we actually mean a few things. This includes both Java-based tests that target an external cluster _and_ YAML tests. We only need the specs for the latter, but we group them together as a single thing. We need to tease these apart.

Not all projects that execute tests against an external cluster (i.e. have a `RestIntegTesttask`) utilize YAML tests. Pulling the copy rest spec stuff out of that task constructor is the first start. The question remains, where do we now put that logic, such that it get's applied to the appropriate projects. Moving it to `RestTestPlugin` actually wouldn't be correct, as this would apply again, to all projects. I suggest we create a new `YamlRestTestPlugin` that does all the spec copying and wiring and we apply this plugin explicitly to projects that leverage YAML tests.